gui/blueprint.lua

```diff
@@ -19,4 +19,159 @@ values set in the interface. See the `blueprint` documentation for information
 on the possible parameters and options.
 ]====]
 
-print('coming soon!')
+local gui = require('gui')
+local guidm = require('gui.dwarfmode')
+local widgets = require('gui.widgets')
+
+local function get_dims(pos1, pos2)
+    local width, height, depth = math.abs(pos1.x - pos2.x) + 1,
+            math.abs(pos1.y - pos2.y) + 1,
+            math.abs(pos1.z - pos2.z) + 1
+    return width, height, depth
+end
+
+ActionPanel = defclass(ActionPanel, widgets.Panel)
+ActionPanel.ATTRS{
+    get_mark_fn=DEFAULT_NIL,
+}
+function ActionPanel:init()
+    self:addviews{
+        widgets.Label{
+            text={{text=self:callback('get_action_text')}},
+            frame={t=0},
+        },
+        widgets.Label{
+            text='with the cursor.',
+            frame={t=1},
+        },
+    }
+end
+function ActionPanel:get_action_text()
+    if self.get_mark_fn() then
+        return 'Select the second corner'
+    end
+    return 'Select the first corner'
+end
+
+BlueprintUI = defclass(BlueprintUI, guidm.MenuOverlay)
+BlueprintUI.ATTRS {
+    presets={},
+    frame_inset=1,
```

Comment:
To allow more specificity in tests (and keybinding contexts, etc.) you could define a `focus_path` field here to cause `getCurFocus()` to return a more detailed string than "dfhack/lua". 

I _think_ this should work (untested):
```suggestion
    frame_inset=1,
    focus_path='blueprint',
```

Example change in tests:

```diff
-expect.eq('dfhack/lua', dfhack.gui.getCurFocus(true))
+expect.eq('dfhack/lua/blueprint', dfhack.gui.getCurFocus(true))
```